README.md
# stlplus-rs

[![Latest Version](https://img.shields.io/crates/v/stlplus-rs.svg)](https://crates.io/crates/stlplus-rs)
[![Documentation](https://docs.rs/stlplus-rs/badge.svg)](https://docs.rs/stlplus-rs)
![ci](https://github.com/nmandery/stlplus-rs/workflows/CI/badge.svg)
[![dependency status](https://deps.rs/repo/github/nmandery/stlplus-rs/status.svg)](https://deps.rs/repo/github/nmandery/stlplus-rs)

Port of the enhanced Seasonal Trend Decomposition using Loess (STL) implementation from
https://github.com/hafen/stlplus (and the java implementation https://github.com/ruananswer/twitter-anomalyDetection-java)
to rust.

At the current stage this project is mostly a rough port of the above repositories, there surely is room for improvement - PRs welcome.

## Example

See [crates_io_downloads.rs](examples/crates_io_downloads.rs).

```rust
cargo run --example crates_io_downloads
```

![](example.png)


## License

<sup>
Licensed under either of <a href="LICENSE-APACHE">Apache License, Version
2.0</a> or <a href="LICENSE-MIT">MIT license</a> at your option.
</sup>


CHANGES.md
# Changes

## Unreleased

## 0.1.3
* Dependency upgrades: criterion 0.3 -> 0.5, itertools 0.10 -> 0.11

## 0.1.2
* Fix for degree offset in t_window calculation.
* Fix porting error from 1-based indexing in R [#3](https://github.com/nmandery/stlplus-rs/pull/3)

## 0.1.1
* Fix out-of-bounds access. [#1](https://github.com/nmandery/stlplus-rs/issues/1)

## 0.1.0

Initial release


src/lib.rs
pub mod error;
mod loess;
mod util;

use crate::error::Error;
use crate::loess::loess_stl;
use crate::util::{NextOddInt, SumAgg, ValidateNotZero};
use itertools::izip;
use num_traits::{AsPrimitive, Float};
use std::fmt::Debug;
pub use util::NextOdd;

#[derive(Debug, Eq, PartialEq, Copy, Clone)]
pub enum Degree {
    Degree0,
    Degree1,
    Degree2,
}

impl Degree {
    /// index in the COEF_XX static arrays
    fn coef_index(&self) -> usize {
        match self {
            Degree::Degree0 => 0,
            Degree::Degree1 => 0,
            Degree::Degree2 => 1,
        }
    }
}

macro_rules! impl_tryfrom_int_for_degree {
    ($($t:ty),*) => {
        $(
            impl TryFrom<$t> for Degree {
                type Error = Error;

                fn try_from(value: $t) -> Result<Self, Self::Error> {
                    match value {
                        1 => Ok(Self::Degree0),
                        2 => Ok(Self::Degree1),
                        3 => Ok(Self::Degree2),
                        _ => Err(Error::InvalidDegree),
                    }
                }
            }
        )*
    };
}
impl_tryfrom_int_for_degree!(i8, i16, i32, i64, i128, isize, u8, u16, u32, u64, u128, usize);

///
/// Comments are still from the original R implementation and Java port.
#[derive(Clone)]
pub struct STLOptions {
    /// The number of observations in each cycle of the seasonal component, n_p
    pub num_obs_per_period: usize,

    /// s.window either the character string \code{"periodic"} or the span (in lags) of the loess window for seasonal extraction,
    /// which should be odd.  This has no default.
    ///
    /// None is used for periodic
    pub s_window: Option<usize>,

    /// s.degree degree of locally-fitted polynomial in seasonal extraction.  Should be 0, 1, or 2.
    pub s_degree: Degree,

    /// t.window the span (in lags) of the loess window for trend extraction, which should be odd.
    /// If \code{NULL}, the default, \code{nextodd(ceiling((1.5*period) / (1-(1.5/s.window))))}, is taken.
    pub t_window: Option<usize>,

    /// t.degree degree of locally-fitted polynomial in trend extraction.  Should be 0, 1, or 2.
    pub t_degree: Degree,

    /// l.window the span (in lags) of the loess window of the low-pass filter used for each subseries.
    ///
    /// Defaults to the smallest odd integer greater than or equal to \code{n.p}
    /// which is recommended since it prevents competition between the trend and seasonal components.
    /// If not an odd integer its given value is increased to the next odd one.
    pub l_window: Option<usize>,

    /// l.degree degree of locally-fitted polynomial for the subseries low-pass filter.  Should be 0, 1, or 2.
    pub l_degree: Degree,

    /// s.jump s.jump,t.jump,l.jump,fc.jump integers at least one to increase speed of the respective smoother.
    /// Linear interpolation happens between every \code{*.jump}th value.
    pub s_jump: Option<usize>,

    /// t.jump
    pub t_jump: Option<usize>,

    /// l.jump
    pub l_jump: Option<usize>,

    /// critfreq the critical frequency to use for automatic calculation of smoothing windows for the trend and high-pass filter.
    pub critfreq: f64,

    /// The number of passes through the inner loop, n_i
    pub number_of_inner_loop_passes: u32,

    /// The number of robustness iterations of the outer loop, n_o
    pub number_of_robustness_iterations: u32,
}

impl Default for STLOptions {
    fn default() -> Self {
        let num_obs_per_period = 4;
        Self {
            num_obs_per_period,
            s_window: None,
            s_degree: Degree::Degree1,
            t_window: None,
            t_degree: Degree::Degree1,
            l_window: None,
            l_degree: Degree::Degree1,
            s_jump: None,
            t_jump: None,
            l_jump: None,
            critfreq: 0.05,
            number_of_inner_loop_passes: 2,
            number_of_robustness_iterations: 1,
            //number_of_data_points: num_obs_per_period * 2,
        }
    }
}

pub struct STLOutput<VALUE> {
    pub trend: Vec<VALUE>,
    pub seasonal: Vec<VALUE>,
    pub remainder: Vec<VALUE>,
}

pub fn stl_decompose<VALUE>(
    values: &[VALUE],
    options: &STLOptions,
) -> Result<STLOutput<VALUE>, Error>
where
    VALUE: Float + 'static + Copy,
    usize: AsPrimitive<VALUE>,
    i64: AsPrimitive<VALUE>,
{
    let n = values.len();
    let times_i64: Vec<_> = (0..n).map(|v| v as i64).collect();

    stl_decompose_with_time(&times_i64, values, options)
}

pub fn stl_decompose_with_time<TIME, VALUE>(
    times: &[TIME],
    values: &[VALUE],
    options: &STLOptions,
) -> Result<STLOutput<VALUE>, Error>
where
    VALUE: Float + 'static + Copy,
    usize: AsPrimitive<VALUE>,
    TIME: AsPrimitive<i64>,
    i64: AsPrimitive<VALUE>,
{
    if values.is_empty() || times.is_empty() {
        return Err(Error::EmptyInputSlice);
    }
    let n = values.len();
    if times.len() != n {
        return Err(Error::InputSlicesDifferingLength);
    }
    if n < (options.num_obs_per_period * 2) {
        return Err(Error::InputSliceToFewObs);
    }

    let times_i64: Vec<i64> = times.iter().map(|t| t.as_()).collect();

    let validated_options = validate_options(options, n)?;
    let mut trend = vec![VALUE::zero(); n];
    let mut seasonal = vec![VALUE::zero(); n];
    let mut deseasonalized = vec![VALUE::zero(); n];

    // cycleSubIndices will keep track of what part of the seasonal each observation belongs to
    let cycle_sub_indices: Vec<_> = (1..=validated_options.num_obs_per_period)
        .cycle()
        .take(values.len())
        .collect();
    let weights = vec![VALUE::one(); n];
    let mut detrend = vec![VALUE::zero(); n];
    // todo: missing stuff from java impl?

    let mut cycle_sub = Vec::with_capacity(
        (n as f64 / validated_options.num_obs_per_period as f64).ceil() as usize / 2,
    );
    let mut sub_weights = Vec::with_capacity(cycle_sub.capacity());

    let (cs1, cs2) = {
        let mut cs1 = Vec::with_capacity(validated_options.num_obs_per_period);
        let mut cs2 = Vec::with_capacity(validated_options.num_obs_per_period);
        for i in 0..validated_options.num_obs_per_period {
            cs1.push(cycle_sub_indices[i]);
            cs2.push(
                cycle_sub_indices[values
                    .len()
                    .saturating_sub(validated_options.num_obs_per_period + i)],
            );
        }
        (cs1, cs2)
    };

    let l_ev = Ev::new(n, validated_options.l_jump);
    let t_ev = Ev::new(n, validated_options.t_jump);

    let mut c = vec![VALUE::nan(); n + 2 * validated_options.num_obs_per_period];

    // start and end indices for after adding in extra n.p before and after
    let c_start_idx = validated_options.num_obs_per_period;
    let c_end_idx = n - 1 + validated_options.num_obs_per_period;

    for _outer_iteration_i in 0..options.number_of_robustness_iterations {
        for _inner_iteration_i in 0..options.number_of_inner_loop_passes {
            // Step 1: detrending
            izip!(detrend.iter_mut(), values.iter(), trend.iter()).for_each(|(dt, v, t)| {
                *dt = *v - *t;
            });

            // step 2: smoothing of cycle-subseries
            for i in 0..validated_options.num_obs_per_period {
                cycle_sub.clear();
                sub_weights.clear();
                let mut j = i;
                while j < n {
                    if cycle_sub_indices[j] == i + 1 {
                        cycle_sub.push(detrend[j]);
                        sub_weights.push(weights[j])
                    }
                    j += validated_options.num_obs_per_period;
                }

                let weight_mean_ans = weight_mean(&cycle_sub, &sub_weights)?;
                j = i;
                while j < validated_options.num_obs_per_period {
                    if cs1[j] == i + 1 {
                        c[j] = weight_mean_ans;
                    }
                    j += validated_options.num_obs_per_period;
                }
                j = i;
                while j < n {
                    if cycle_sub_indices[j] == i + 1 {
                        c[j + validated_options.num_obs_per_period] = weight_mean_ans;
                    }
                    j += validated_options.num_obs_per_period;
                }
                for j in 0..validated_options.num_obs_per_period {
                    if cs2[j] == i + 1 {
                        c[j + validated_options.num_obs_per_period + n] = weight_mean_ans;
                    }
                }
            }

            // Step 3: Low-pass filtering of collection of all the cycle-subseries
            // moving averages
            let ma3 = cycle_subseries_moving_averages(&c, validated_options.num_obs_per_period);

            // Step 4: Detrend smoothed cycle-subseries
            let l = loess_stl(
                &times_i64,
                &ma3,
                validated_options.l_window,
                validated_options.l_degree,
                l_ev.as_slice(),
                &weights,
                validated_options.l_jump,
            )?;

            // Step 5: Deseasonalize
            izip!(
                seasonal.iter_mut(),
                (&c)[c_start_idx..=c_end_idx].iter(),
                l.iter(),
                values.iter(),
                deseasonalized.iter_mut()
            )
            .for_each(|(s, c, l, v, d)| {
                *s = *c - *l;
                *d = *v - *s;
            });

            // Step 6: Trend Smoothing
            trend = loess_stl(
                &times_i64,
                &deseasonalized,
                validated_options.t_window,
                validated_options.t_degree,
                t_ev.as_slice(),
                &weights,
                validated_options.t_jump,
            )?;
        }
    }

    let remainder: Vec<_> = izip!(values.iter(), trend.iter(), seasonal.iter())
        .map(|(v, t, s)| *v - *t - *s)
        .collect();

    Ok(STLOutput {
        trend,
        seasonal,
        remainder,
    })
}

struct Ev {
    n: usize,
    array_min_len: usize,
    storage_vec: Vec<usize>,
}

impl Ev {
    fn new(n: usize, jump: usize) -> Self {
        let array_min_len = (n as f64 / jump as f64).ceil() as usize;
        let mut storage_vec = vec![0usize; array_min_len + 1];

        let mut i = 0;
        let mut j = 0;
        while i < array_min_len {
            storage_vec[i] = j;
            i += 1;
            j += jump
        }

        // always have the last element == (n - 1)
        storage_vec[array_min_len] = n - 1;

        Self {
            n,
            array_min_len,
            storage_vec,
        }
    }

    /// return a slice where the last element == `n`
    fn as_slice(&self) -> &[usize] {
        if self.storage_vec[self.array_min_len - 1] != self.n - 1 {
            &self.storage_vec
        } else {
            &self.storage_vec[0..self.array_min_len]
        }
    }
}

struct ValidatedOptions {
    num_obs_per_period: usize,
    //number_of_data_points: usize,
    //s_window: usize,
    //s_degree: Degree,
    //s_jump: usize,
    t_window: usize,
    t_degree: Degree,
    t_jump: usize,
    l_window: usize,
    l_degree: Degree,
    l_jump: usize,
    //periodic: bool,
}

fn validate_options(options: &STLOptions, num_values: usize) -> Result<ValidatedOptions, Error> {
    let num_obs_per_period = if options.num_obs_per_period >= 4 {
        options.num_obs_per_period
    } else {
        return Err(Error::InvalidNumObsPerPeriod);
    };

    /*
    let number_of_data_points = if options.number_of_data_points > 2 * num_obs_per_period {
        options.number_of_data_points
    } else {
        return Err(Error::InvalidNumberOfDataPoints);
    };

     */

    let l_degree = options.l_degree;
    let l_window = options.l_window.unwrap_or(num_obs_per_period).next_odd();
    let l_jump = options.l_jump.unwrap_or_else(|| window_to_jump(l_window));

    let (s_window, s_degree, _s_jump, _periodic) = if let Some(s_window) = options.s_window {
        let s_window = validate_window(s_window)?;
        let s_jump = options.s_jump.unwrap_or_else(|| window_to_jump(s_window));
        (s_window, options.s_degree, s_jump, false)
    } else {
        // periodic
        let s_window = 10 * num_values + 1;
        let s_degree = Degree::Degree0;
        let s_jump = window_to_jump(s_window);
        (s_window, s_degree, s_jump, true)
    };

    let t_degree = options.t_degree;
    let t_window = if let Some(t_window) = options.t_window {
        validate_window(t_window)?
    } else {
        get_t_window(
            t_degree,
            s_degree,
            s_window,
            num_obs_per_period,
            options.critfreq,
        )?
    };
    let t_jump = options.t_jump.unwrap_or_else(|| window_to_jump(t_window));

    Ok(ValidatedOptions {
        num_obs_per_period,
        //number_of_data_points,
        //s_window,
        //s_degree,
        //s_jump,
        t_window,
        t_degree,
        t_jump,
        l_window,
        l_degree,
        l_jump,
        //periodic,
    })
}

static COEFS_A: [[f64; 2]; 2] = [
    [0.000103350651767650, 3.81086166990428e-6],
    [-0.000216653946625270, 0.000708495976681902],
];
static COEFS_B: [[f64; 2]; 3] = [
    [1.42686036792937, 2.24089552678906],
    [-3.1503819836694, -3.30435316073732],
    [5.07481807116087, 5.08099438760489],
];
static COEFS_C: [[f64; 2]; 3] = [
    [1.66534145060448, 2.33114333880815],
    [-3.87719398039131, -1.8314816166323],
    [6.46952900183769, 1.85431548427732],
];

fn get_t_window(
    t_degree: Degree,
    s_degree: Degree,
    s_window: usize,
    num_obs_per_period: usize,
    critfreq: f64,
) -> Result<usize, Error> {
    let s_index = s_degree.coef_index();
    let t_index = t_degree.coef_index();

    // estimate critical frequency for seasonal
    let betac0 = COEFS_A[1][s_index].mul_add(critfreq, COEFS_A[0][s_index]);
    let betac1 = COEFS_B[1][s_index].mul_add(critfreq, COEFS_B[0][s_index])
        + COEFS_B[2][s_index] * critfreq.powi(2);
    let betac2 = COEFS_C[1][s_index].mul_add(critfreq, COEFS_C[0][s_index])
        + COEFS_C[2][s_index] * critfreq.powi(2);

    let f_c = (1.0 - (betac0 + betac1 / s_window as f64 + betac2 / s_window.pow(2) as f64))
        / num_obs_per_period as f64;

    // choose
    let betat0 = COEFS_A[1][t_index].mul_add(critfreq, COEFS_A[0][t_index]);
    let betat1 = COEFS_B[1][t_index].mul_add(critfreq, COEFS_B[0][t_index])
        + COEFS_B[2][t_index] * critfreq.powi(2);
    let betat2 = COEFS_C[1][t_index].mul_add(critfreq, COEFS_C[0][t_index])
        + COEFS_C[2][t_index] * critfreq.powi(2);

    let betat00 = betat0 - f_c;

    Ok(
        ((-betat1 - (betat1.powi(2) - 4.0 * betat00 * betat2).sqrt()) / (2.0 * betat00)).next_odd()
            as usize,
    )
}

fn validate_window(window: usize) -> Result<usize, Error> {
    if window < 1 {
        Err(Error::InvalidWindow)
    } else {
        Ok(window)
    }
}

fn window_to_jump(window: usize) -> usize {
    (window as f64 / 10.0).ceil() as usize
}

/// cycle-subseries moving averages
///
/// `num_obs_per_period` is the periodicity `n_p`
///
/// This function was called `c_ma` in stlplus
pub(crate) fn cycle_subseries_moving_averages<F>(x: &[F], num_obs_per_period: usize) -> Vec<F>
where
    F: Float + 'static + Copy,
    usize: AsPrimitive<F>,
{
    let nn = x.len().saturating_sub(num_obs_per_period * 2);

    let mut ans = vec![F::zero(); x.len() - 2 * num_obs_per_period];
    let mut ma = vec![F::zero(); nn + num_obs_per_period + 1];
    let mut ma2 = vec![F::zero(); nn + 2];

    let mut ma_tmp = x[0..num_obs_per_period].sum_agg();
    ma[0] = ma_tmp / num_obs_per_period.as_();

    for i in num_obs_per_period..(nn + 2 * num_obs_per_period) {
        ma_tmp = ma_tmp - x[i - num_obs_per_period] + x[i];
        ma[i - num_obs_per_period + 1] = ma_tmp / num_obs_per_period.as_();
    }

    ma_tmp = (ma[0..num_obs_per_period]).sum_agg();
    ma2[0] = ma_tmp / num_obs_per_period.as_();
    for i in num_obs_per_period..(nn + num_obs_per_period + 1) {
        ma_tmp = ma_tmp - ma[i - num_obs_per_period] + ma[i];
        ma2[i - num_obs_per_period + 1] = ma_tmp / num_obs_per_period.as_();
    }

    ma_tmp = (ma2[0..3]).sum_agg();
    ans[0] = ma_tmp / 3usize.as_();
    for i in 3..(nn + 2) {
        ma_tmp = ma_tmp - ma2[i - 3] + ma2[i];
        ans[i - 2] = ma_tmp / 3usize.as_();
    }

    ans
}

fn weight_mean<T>(x: &[T], w: &[T]) -> Result<T, Error>
where
    T: Float,
{
    let (sum, sum_w) = x.iter().zip(w.iter()).fold(
        (T::zero(), T::zero()),
        |(sum, sum_w), (x_value, w_value)| {
            if !x_value.is_nan() {
                (sum + (*x_value * *w_value), sum_w + *w_value)
            } else {
                (sum, sum_w)
            }
        },
    );
    Ok(sum / sum_w.validate_not_zero()?)
}

#[cfg(test)]
mod tests {
    use crate::{stl_decompose, STLOptions};

    #[test]
    fn c_ma() {
        let input = vec![1.0f32, 1.0, 2.0, 2.0, 3.0, 3.0, 2.0, 2.0, 1.0, 1.0];
        let n_p = 3;
        let out = super::cycle_subseries_moving_averages(&input, n_p);
        dbg!(out);
    }

    /// https://github.com/nmandery/stlplus-rs/issues/1
    #[test]
    fn indexing_within_bounds() {
        let input = vec![0.0f32; 2581];
        let options = STLOptions {
            num_obs_per_period: 365,
            ..Default::default()
        };
        // should not cause out-of-bounds panic
        stl_decompose(&input, &options).unwrap();
    }

    /// https://github.com/nmandery/stlplus-rs/issues/1
    #[test]
    fn input_too_short() {
        let input = vec![0.0f32; 25];
        let options = STLOptions {
            num_obs_per_period: 365,
            ..Default::default()
        };
        // should not cause out-of-bounds panic
        assert!(stl_decompose(&input, &options).is_err());
    }
}


src/util.rs
use crate::Error;
use num_traits::{AsPrimitive, Float, Zero};
use std::ops::Add;

pub trait NextOddInt {
    fn next_odd(&self) -> Self;
}

macro_rules! impl_nextoddint {
    ($($t:ty),*) => {
        $(
            impl NextOddInt for $t {
                fn next_odd(&self) -> Self {
                    if self % 2 == 0 {
                        self + 1
                    } else {
                        *self
                    }
                }
            }
        )*
    };
}
impl_nextoddint!(i8, i16, i32, i64, i128, isize, u8, u16, u32, u64, u128, usize);

pub trait NextOdd {
    fn next_odd(&self) -> i64;
}

impl<T> NextOdd for T
where
    T: Float + 'static + Copy + AsPrimitive<i64>,
{
    fn next_odd(&self) -> i64 {
        let rounded: i64 = self.round().as_();
        rounded.next_odd()
    }
}

pub trait ValidateNotZero {
    fn validate_not_zero(self) -> Result<Self, Error>
    where
        Self: Sized;
}

impl<T> ValidateNotZero for T
where
    T: Zero,
{
    fn validate_not_zero(self) -> Result<Self, Error> {
        if self.is_zero() {
            Err(Error::UnexpectedZero)
        } else {
            Ok(self)
        }
    }
}

pub trait SumAgg {
    type Output;

    fn sum_agg(&self) -> Self::Output;
}

impl<T> SumAgg for [T]
where
    T: Zero + Copy + Add,
{
    type Output = T;

    fn sum_agg(&self) -> Self::Output {
        self.iter().fold(T::zero(), |a, b| a + *b)
    }
}


src/arrow.rs
use crate::{Error, STLOptions, STLOutput};
use arrow2::array::{Array, PrimitiveArray};
use arrow2::types::NativeType;
use num_traits::{AsPrimitive, Float};

pub struct ArrowDecomposeOutput<VALUE: NativeType> {
    pub trend: PrimitiveArray<VALUE>,
    pub seasonal: PrimitiveArray<VALUE>,
    pub remainder: PrimitiveArray<VALUE>,
}

impl<VALUE> From<STLOutput<VALUE>> for ArrowDecomposeOutput<VALUE>
where
    VALUE: NativeType + Float,
{
    fn from(decomp_output: STLOutput<VALUE>) -> Self {
        Self {
            trend: to_primitive_array(decomp_output.trend),
            seasonal: to_primitive_array(decomp_output.seasonal),
            remainder: to_primitive_array(decomp_output.remainder),
        }
    }
}

pub fn stl_decompose_arrow<VALUE>(
    values: PrimitiveArray<VALUE>,
    options: &STLOptions,
) -> Result<ArrowDecomposeOutput<VALUE>, Error>
where
    VALUE: Float + NativeType + 'static + Copy,
    usize: AsPrimitive<VALUE>,
    i64: AsPrimitive<VALUE>,
{
    let values_vec: Vec<_> = (0..values.len()).map(|i| get_or_nan(&values, i)).collect();
    super::stl_decompose(&values_vec, options).map(|decomp_out| decomp_out.into())
}

pub fn stl_decompose_arrow_with_time<TIME, VALUE>(
    times: PrimitiveArray<TIME>,
    values: PrimitiveArray<VALUE>,
    options: &STLOptions,
) -> Result<ArrowDecomposeOutput<VALUE>, Error>
where
    VALUE: Float + NativeType + 'static + Copy,
    TIME: NativeType + AsPrimitive<i64>,
    usize: AsPrimitive<VALUE>,
    i64: AsPrimitive<VALUE>,
{
    // all times must be set, invalid values are not allowed
    if !all_values_valid(&times) {
        return Err(Error::InputSliceInvalidValues);
    }

    let values_vec: Vec<_> = (0..values.len()).map(|i| get_or_nan(&values, i)).collect();
    let times: Vec<_> = (0..times.len()).map(|i| times.value(i)).collect();
    super::stl_decompose_with_time(&times, &values_vec, options).map(|decomp_out| decomp_out.into())
}

#[inline(always)]
fn get_or_nan<T>(pa: &PrimitiveArray<T>, position: usize) -> T
where
    T: Float + NativeType,
{
    if pa.is_valid(position) {
        pa.value(position)
    } else {
        T::nan()
    }
}

fn to_primitive_array<T>(mut contents: Vec<T>) -> PrimitiveArray<T>
where
    T: Float + NativeType,
{
    PrimitiveArray::from_iter(
        contents
            .drain(..)
            .map(|v| if v.is_nan() { None } else { Some(v) }),
    )
}

fn all_values_valid<T>(array: &PrimitiveArray<T>) -> bool
where
    T: NativeType,
{
    array
        .validity()
        .map(|validity| validity.null_count())
        .unwrap_or(0)
        == 0
}

#[cfg(test)]
mod tests {
    use arrow2::array::PrimitiveArray;

    #[test]
    fn all_values_valid() {
        let pa1 = PrimitiveArray::<i32>::from([Some(1), Some(2), Some(3)]);
        assert!(super::all_values_valid(&pa1));

        let pa2 = PrimitiveArray::<i32>::from([Some(1), None, Some(3)]);
        assert!(!super::all_values_valid(&pa2));
    }
}


src/error.rs
#[derive(thiserror::Error, Debug)]
pub enum Error {
    #[error("empty input slice given")]
    EmptyInputSlice,

    #[error("input slice of differing length")]
    InputSlicesDifferingLength,

    #[error("input slice contains invalid values")]
    InputSliceInvalidValues,

    #[error("Invalid degree - must be 1, 2 or 3")]
    InvalidDegree,

    #[error("Invalid window - must be > 0")]
    InvalidWindow,

    #[error("length of sub_labels must match num_obs_per_period")]
    InvalidSubLabelsLength,

    #[error("sub_labels must be unique")]
    SubLabelsMustBeUnique,

    #[error("num_obs_per_period is invalid")]
    InvalidNumObsPerPeriod,
    // InvalidNumberOfDataPoints,
    #[error("unexpected 0 encountered")]
    UnexpectedZero,

    #[error("input slice contains to few observations - must be more than two periods")]
    InputSliceToFewObs,
}


benches/simple_stl.rs
use criterion::{criterion_group, criterion_main, Criterion};
use std::time::Duration;
use stlplus_rs::{stl_decompose, STLOptions};

#[rustfmt::skip]
fn daily_downloads() -> Vec<f32> {
    // see fetch_crates_io_testdata.py
    vec![
        108113.0, 103580.0, 85339.0, 40711.0, 45263.0, 84316.0, 108669.0, 110202.0, 106642.0,
        105171.0, 43743.0, 45082.0, 110190.0, 110791.0, 107311.0, 109093.0, 99330.0, 43257.0,
        47701.0, 113499.0, 111596.0, 104487.0, 103415.0, 109459.0, 47813.0, 42045.0, 118689.0,
        114456.0, 112742.0, 108783.0, 89554.0, 39153.0, 41106.0, 101895.0, 102104.0, 105788.0,
        110034.0, 107436.0, 47510.0, 43332.0, 112624.0, 116730.0, 109939.0, 94388.0, 90817.0,
        41149.0, 39748.0, 80703.0, 103159.0, 115283.0, 102565.0, 101867.0, 42052.0, 41680.0,
        111712.0, 119406.0, 112199.0, 120700.0, 109765.0, 46288.0, 41949.0, 126540.0, 128660.0,
        119960.0, 116728.0, 84615.0, 41455.0, 32614.0, 86439.0, 95445.0, 92745.0, 92750.0, 84713.0,
        38181.0, 35594.0, 92890.0, 98724.0, 88924.0, 87921.0, 91981.0, 30671.0, 29672.0, 70103.0,
        83425.0, 89964.0, 88415.0, 76165.0, 30552.0, 25771.0, 43555.0,
    ]
}

fn criterion_benchmark(c: &mut Criterion) {
    let mut group = c.benchmark_group("simple_stl");
    group.sample_size(1000);
    group.warm_up_time(Duration::from_secs(2));
    group.bench_function("simple STL", |bencher| {
        let data = daily_downloads();

        let options = STLOptions {
            num_obs_per_period: 7,
            ..Default::default()
        };
        bencher.iter(|| {
            let _decomp = stl_decompose(&data, &options).unwrap();
        });
    });
    group.finish();
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);


examples/fetch_crates_io_testdata.py
#!/usr/bin/env python
#
# Create some simple test-data by downloading crates.io download-stats. These provide some nice
# weekly seasonality and trends as new versions are released.

import requests

crate_name = "hashbrown"

d = requests.get(f"https://crates.io/api/v1/crates/{crate_name}/versions").json()
version_by_id = {}
for c in d["versions"]:
    version_by_id[c["id"]] = c["num"]

d = requests.get(f"https://crates.io/api/v1/crates/{crate_name}/downloads").json()
dl_by_id = {}
for dl in d["version_downloads"]:
    if not dl["version"] in dl_by_id:
        dl_by_id[dl["version"]] = []
    dl_by_id[dl["version"]].append((dl["date"], float(dl.get("downloads", 0))))
    
for k, v in dl_by_id.items():
    sorted(v, key=lambda x: x[0])
    print(f"\n\n{crate_name} version {version_by_id[k]}\ndownloads: {', '.join([str(x[1]) for x in v])}")


examples/crates_io_downloads.rs
use num_traits::Float;
use plotters::prelude::*;
use stlplus_rs::{stl_decompose, STLOptions, STLOutput};

#[rustfmt::skip]
fn daily_downloads() -> Vec<f32> {
    // see fetch_crates_io_testdata.py
    vec![
        108113.0, 103580.0, 85339.0, 40711.0, 45263.0, 84316.0, 108669.0, 110202.0, 106642.0,
        105171.0, 43743.0, 45082.0, 110190.0, 110791.0, 107311.0, 109093.0, 99330.0, 43257.0,
        47701.0, 113499.0, 111596.0, 104487.0, 103415.0, 109459.0, 47813.0, 42045.0, 118689.0,
        114456.0, 112742.0, 108783.0, 89554.0, 39153.0, 41106.0, 101895.0, 102104.0, 105788.0,
        //f32::NAN,
        110034.0, 107436.0, 47510.0, 43332.0, 112624.0, 116730.0, 109939.0, 94388.0, 90817.0,
        41149.0, 39748.0, 80703.0, 103159.0, 115283.0, 102565.0, 101867.0, 42052.0, 41680.0,
        111712.0, 119406.0, 112199.0, 120700.0, 109765.0, 46288.0, 41949.0, 126540.0, 128660.0,
        119960.0, 116728.0, 84615.0, 41455.0, 32614.0, 86439.0, 95445.0, 92745.0, 92750.0, 84713.0,
        38181.0, 35594.0, 92890.0, 98724.0, 88924.0, 87921.0, 91981.0, 30671.0, 29672.0, 70103.0,
        83425.0, 89964.0, 88415.0, 76165.0, 30552.0, 25771.0, 43555.0,
    ]
}

const OUT_FILE_NAME: &str = "downloads.png";

fn main() {
    let data = daily_downloads();

    let options = STLOptions {
        num_obs_per_period: 7,
        ..Default::default()
    };
    let decomp = stl_decompose(&data, &options).unwrap();

    plot(&data, &decomp, OUT_FILE_NAME);
}

fn plot<A>(input: &[f32], decomp: &STLOutput<f32>, out_file: A)
where
    A: AsRef<std::path::Path>,
{
    let root = BitMapBackend::new(OUT_FILE_NAME, (1000, 1000)).into_drawing_area();
    root.fill(&WHITE).unwrap();
    let root = root.margin(10, 10, 10, 10);
    let drawing_areas = root.split_evenly((4, 1));

    let (_min_value, max_value) = min_max(input);
    let (min_value, _max_value) = min_max(&decomp.seasonal);

    let sub_plots = vec![
        ("Input data", &drawing_areas[0], &BLACK, input),
        ("Seasonality", &drawing_areas[1], &BLUE, &decomp.seasonal),
        ("Trend", &drawing_areas[2], &BLUE, &decomp.trend),
        ("Remainder", &drawing_areas[3], &BLUE, &decomp.remainder),
    ];

    for (title, drawing_area, color, data_slice) in sub_plots.iter() {
        let mut chart = ChartBuilder::on(drawing_area)
            .set_label_area_size(LabelAreaPosition::Left, 80)
            .set_label_area_size(LabelAreaPosition::Bottom, 30)
            .build_cartesian_2d(0.0..(data_slice.len() as f32), min_value..max_value)
            .unwrap();

        chart
            .configure_mesh()
            .disable_mesh()
            .y_desc(*title)
            .x_desc("Time step")
            .draw()
            .unwrap();

        // plot all segments without a nan value
        let mut i = 0;
        let mut j = 0;
        while j + 1 < data_slice.len() {
            while i < data_slice.len() && data_slice[i].is_nan() {
                i += 1;
            }
            j = i;
            while j < data_slice.len() && !data_slice[j].is_nan() {
                j += 1;
            }
            if i != j {
                chart
                    .draw_series(LineSeries::new(
                        data_slice[i..j]
                            .iter()
                            .enumerate()
                            .map(|(idx, value)| ((idx + i) as f32, *value)),
                        color,
                    ))
                    .unwrap();
            }
            i = j;
        }
    }

    root.present().unwrap();
    println!(
        "Plot has been saved to {}",
        out_file.as_ref().to_str().unwrap()
    );
}

fn min_max<T>(slice: &[T]) -> (T, T)
where
    T: Float,
{
    let (min_value, max_value) = slice.iter().fold(
        (*slice.first().unwrap(), *slice.first().unwrap()),
        |acc, value| {
            (
                if *value < acc.0 { *value } else { acc.0 },
                if *value > acc.1 { *value } else { acc.1 },
            )
        },
    );
    (min_value, max_value)
}
